File: problems/data-structures-n-algorithms/wip/22.cpp
// 22. Generate Parentheses
// https://leetcode.com/problems/generate-parentheses/

#include <bits/stdc++.h>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <vector>

using namespace std;

vector<string> copyStrs(const string &str, vector<string> &arr) {
	vector<string> res(arr.size());
	int i;

	for (i = 0; i < arr.size(); i++) {
		res[i] = str + arr[i];
	}

	return res;
}

vector<string> generate(int left, int right) {
	if (right == 0) {
		vector<string> arr(1);
		arr[0] = string(")");
		return arr;
	}

	vector<string> s1, s2;

	if (left <= right && left != 0) {
		string curr;
		curr.append(string("("));
		s1 = generate(left - 1, right);
		s1 = copyStrs(curr, s1);
	}
	if (right > left) {
		string curr;
		curr.append(string(")"));
		s2 = generate(left, right - 1);
		s2 = copyStrs(curr, s2);
	}

	vector<string> res;
	res.reserve(s1.size() + s2.size());
	res.insert(res.end(), s2.begin(), s2.end());
	res.insert(res.end(), s2.begin(), s2.end());

	return res;
}

vector<string> generateParenthesis(int n) {
	int left = n, right = n;
	vector<string> res;

	if (n == 0) {
		return res;
	}

	return generate(left, right);
}

int main() {
	int n, i;

	cin >> n;

	vector<string> strs = generateParenthesis(n);

	for (i = 0; i < n; i++) {
		cout << strs[i];
	}
}

File: problems/data-structures-n-algorithms/medium/22.java
// 22. Generate Parentheses
// https://leetcode.com/problems/generate-parentheses/

import java.util.ArrayList;
import java.util.List;

class S22 {
	public static void main (String args[]) {
		int n = 3;
		List<String> res = generateParenthesis(n);
		System.out.println(res);
	}

	public static List<String> generateParenthesis (int n) {
		List<String> res = new ArrayList<>();
		recursion(0, 0, 0, n, new StringBuilder(), res);
		return res;
	}

	public static void recursion (int begin, int end, int idx, int n, StringBuilder sb, List<String> res) {
		if (idx == n * 2) {
			res.add(sb.toString());
			return;
		}
		if (begin < n) {
			sb.append('(');
			recursion(begin + 1, end, idx + 1, n, sb, res);
			sb.deleteCharAt(idx);
		}
		if (begin > end) {
			sb.append(')');
			recursion(begin, end + 1, idx + 1, n, sb, res);
			sb.deleteCharAt(idx);
		}
	}
}